src/java.base/share/classes/sun/security/provider/HSS.java

```diff
@@ -0,0 +1,744 @@
+        @java.io.Serial
+        private static final long serialVersionUID = 21L;
+        final int type;
+        final int otsType;
+        final transient LMParams lmParams;
+        private final byte[] I;
+        private final byte[] T1;
+
+        public static LMSPublicKey of(byte[] keyArray) throws InvalidKeyException {
+            return new LMSPublicKey(keyArray, 0, true);
+        }
+
+        public LMSPublicKey(byte[] keyArray, int offset, boolean checkExactLength) throws InvalidKeyException {
+            int inLen = keyArray.length - offset;
+            if (inLen < 8)
+                throw new InvalidKeyException("LMS public key is too short");
+            type = LMSUtils.fourBytesToInt(keyArray, offset);
+            otsType = LMSUtils.fourBytesToInt(keyArray, offset + 4);
+            LMOTSParams lmotsParams;
+
+            try {
+                lmParams = new LMParams(type);
+                lmotsParams = LMOTSParams.of(otsType);
+            } catch (IllegalArgumentException e) {
+                throw new InvalidKeyException(e.getMessage());
+            }
+
+            int m = lmParams.m;
+            if ((inLen < (24 + m)) || (checkExactLength && (inLen != (24 + m))) ||
+                    !lmotsParams.hashAlgName.equals(lmParams.hashAlgStr) ||
+                    (lmParams.m != lmotsParams.n)) {
+                throw new InvalidKeyException("Wrong LMS public Key length");
+            }
+
+            I = Arrays.copyOfRange(keyArray, offset + 8, offset + 8 + 16);
+            T1 = Arrays.copyOfRange(keyArray, offset + 24, offset + 24 + m);
+        }
+
+        public void getI(byte[] arr, int pos) {
+            System.arraycopy(I, 0, arr, pos, 16);
+        }
+
+        public boolean isT1(byte[] arr, int pos) {
+            int m = lmParams.m;
+            int diff = 0;
+            for (int i = 0; i < m; i++) {
+                diff |= (T1[i] ^ arr[pos + i]);
+            }
+            return (diff == 0);
+        }
+
+        public byte[] keyArray() {
+            byte[] result = new byte[keyArrayLength()];
+            LMSUtils.intToFourBytes(type, result, 0);
+            LMSUtils.intToFourBytes(otsType, result, 4);
+            System.arraycopy(I, 0, result, 8, 16);
+            System.arraycopy(T1, 0, result, 24, lmParams.m);
+            return result;
+        }
+
+        public int keyArrayLength() {
+            return 24 + lmParams.m;
+        }
+
+        public String getDigestAlgorithm() {
+            return lmParams.hashAlgStr;
+        }
+    }
+
+    static class LMSUtils {
+        public final static int LMS_RESERVED = 0;
+        public final static int LMS_SHA256_M32_H5 = 5;
+        public final static int LMS_SHA256_M32_H10 = 6;
+        public final static int LMS_SHA256_M32_H15 = 7;
+        public final static int LMS_SHA256_M32_H20 = 8;
+        public final static int LMS_SHA256_M32_H25 = 9;
+        public final static int LMS_SHA256_M24_H5 = 10;
```

Comment:
I think unused code is a distraction. That said, SHA256_M24 used to work. Can't we be better than BC and support M24?

Reply:
I have removed everything that is not SHA-256, but kept a version of the code that has all those removed things so that in case we decide to put it back, it will be there in my work tree.